### PR TITLE
Check for MSBuild property for RestoreProjectStyle on legacy project systems

### PR DESCRIPTION
**Pull request by sbanni**
@emgarten @jainaashish

### CODE REVIEW

File: src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs [hunk 1]
```diff
@@ -21,6 +21,9 @@ namespace NuGet.PackageManagement.VisualStudio
     /// </summary>
     public class EnvDTEProjectAdapter : IEnvDTEProjectAdapter
     {
+        private const string RestoreProjectStyle = "RestoreProjectStyle";
+        private const string ProjectStylePackageReference = "PackageReference";
```
Comment (emgarten) [inline src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs]: Use ProjectStyle.PackageReference instead

File: src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs [hunk 2]
```diff
@@ -156,10 +159,15 @@ public bool IsLegacyCSProjPackageReferenceProject
 
                 if (!_isLegacyCSProjPackageReferenceProject.HasValue)
                 {
+                    var restoreProjectStyle = GetMSBuildProperty(AsIVsBuildPropertyStorage, RestoreProjectStyle);
+                    if (!string.IsNullOrWhiteSpace(restoreProjectStyle) && restoreProjectStyle.Equals(ProjectStylePackageReference, StringComparison.OrdinalIgnoreCase))
```
Comment (emgarten) [inline src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs]: The property check looks good, but the property alone isn't enough. The project must be a VSProject4. It looks like you could combine this with the other check below to also verify that RestoreProjectStyle is not set to PackageReference.

**Review by emgarten — CHANGES_REQUESTED** [review]
A couple minor changes, overall it looks like this would solve it

**Review by jainaashish — CHANGES_REQUESTED** [review]
I don't understand the reason for this PR. We reverted this change to fix RPS perf regression but this PR will again introduce that regression. Ideally we need to come up with a solution which doesn't impact perf for normal scenarios.

File: src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs [hunk 3]
```diff
@@ -159,18 +162,23 @@ public bool IsLegacyCSProjPackageReferenceProject
...
                     {
                         _isLegacyCSProjPackageReferenceProject = false;
                     }
                     else
                     {
-                        _isLegacyCSProjPackageReferenceProject = true;
+                        var restoreProjectStyle = GetMSBuildProperty(AsIVsBuildPropertyStorage, RestoreProjectStyle, null);
+
+                        if (!restoreProjectStyle?.Equals(ProjectStyle.PackageReference.ToString(), StringComparison.OrdinalIgnoreCase) ?? true
+                            && (AsVSProject4.PackageReferences?.InstalledPackages?.Length ?? 0) == 0)
```
Comment (jainaashish) [inline src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs:173]: These two conditions can't be clubbed together. Justin just meant `AsVSProject4 == null` check even if `RestoreProjectStyle` is set to `PackageReference`.

Scenarios are like:
1. If `AsVSProject4 == null` then simply set _isLegacyCSProjPackageReferenceProject = false
2. Else if, `RestoreProjectStyle is defined to PackageReference OR  AsVSProject4.PackageReferences?.InstalledPackages?.Length ?? 0) > 0` then set  _isLegacyCSProjPackageReferenceProject = true
3. Else, set _isLegacyCSProjPackageReferenceProject = false

File: src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs [hunk 4]
```diff
@@ -159,18 +162,23 @@ public bool IsLegacyCSProjPackageReferenceProject
...
-                    else if (AsVSProject4 == null ||
-                        (AsVSProject4.PackageReferences?.InstalledPackages?.Length ?? 0) == 0)
+                    if (AsIVsHierarchy?.IsCapabilityMatch("CPS") ?? true || AsVSProject4 == null)
                     {
                         _isLegacyCSProjPackageReferenceProject = false;
                     }
                     else
                     {
-                        _isLegacyCSProjPackageReferenceProject = true;
+                        var restoreProjectStyle = GetMSBuildProperty(AsIVsBuildPropertyStorage, RestoreProjectStyle, null);
```
Comment (jainaashish) [inline src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs:170]: 👍

File: src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs (hunk 3 above)
Comment (sbanni) [reply src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs:173]: 1 is covered above: if (AsIVsHierarchy?.IsCapabilityMatch("CPS") ?? true || AsVSProject4 == null)
Do you have specific tests you would like me to run to verify that everything works correctly?

File: src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs (hunk 1 above)
Comment (jainaashish) [inline src/NuGet.Clients/PackageManagement.VisualStudio/EnvDTEProjectAdapter.cs]: remove this const `ProjectStylePackageReference `

**Review by jainaashish — APPROVED** [review]
Looks good 👍

**Review by emgarten — APPROVED** [review]

**Comment by emgarten** [comment]
Replaced by @jainaashish's PR

**Pull request closed without merging**